diff -urb zkc-3.4.14/c/src/zookeeper.c buffer-overflow/c/src/zookeeper.c
--- zkc-3.4.14/c/src/zookeeper.c	2019-03-06 11:50:47.000000000 -0500
+++ buffer-overflow/c/src/zookeeper.c	2022-01-04 10:05:28.000000000 -0500
@@ -3477,7 +3477,7 @@
 
 static const char* format_endpoint_info(const struct sockaddr_storage* ep)
 {
-    static char buf[128];
+    static char buf[128+6];
     char addrstr[128];
     void *inaddr;
 #ifdef WIN32
